cpp/tests/bitmask/bit_mask_test.cu

```diff
@@ -0,0 +1,436 @@
+  }
+  
+  gdf_size_type local_count = 0;
+  EXPECT_EQ(GDF_SUCCESS, count_bits(&local_count, bit_mask));
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::destroy_bit_mask(bits));
+
+  EXPECT_EQ(128U, local_count);
+}
+
+TEST_F(BitMaskTest, NotMultipleOfEight)
+{
+  const int num_rows = 1023;
+
+  bit_mask_t *bits = nullptr;
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::create_bit_mask(&bits, num_rows, 0));
+
+  BitMask bit_mask(bits, num_rows);
+
+  for (int i = 7 ; i < num_rows ; i += 8) {
+    set_bit<<<1,1>>>(i, bit_mask);
+  }
+  
+  gdf_size_type local_count = 0;
+  EXPECT_EQ(GDF_SUCCESS, count_bits(&local_count, bit_mask));
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::destroy_bit_mask(bits));
+
+  EXPECT_EQ(127U, local_count);
+}
+
+TEST_F(BitMaskTest, TenThousandRows)
+{
+  const int num_rows = 10000;
+
+  bit_mask_t *bits = nullptr;
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::create_bit_mask(&bits, num_rows, 1));
+
+  BitMask bit_mask(bits, num_rows);
+
+  gdf_size_type local_count = 0;
+  EXPECT_EQ(GDF_SUCCESS, count_bits(&local_count, bit_mask));
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::destroy_bit_mask(bits));
+
+  EXPECT_EQ(10000U, local_count);
+}
+
+TEST_F(BitMaskTest, PerformanceTest)
+{
+  const int num_rows = 100000000;
+
+  bit_mask_t *bits = nullptr;
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::create_bit_mask(&bits, num_rows, 0));
+
+  BitMask bit_mask(bits, num_rows);
+
+  int num_elements = bit_mask::num_elements(num_rows);
+  int block_size = 256;
+  int grid_size = (num_elements + block_size - 1)/block_size;
+
+  uint32_t *local_valid = (uint32_t *) malloc(num_elements * sizeof(uint32_t));
+  for (int i = 0 ; i < num_elements ; ++i) {
+    local_valid[i] = 0x55555555U;
+  }
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::copy_bit_mask(bit_mask.get_valid(), local_valid, num_rows, cudaMemcpyHostToDevice));
+
+  auto start = std::chrono::system_clock::now();
+  cudaProfilerStart();
+  for(int i = 0; i < 1000; ++i) {
+    gdf_size_type local_count = 0;
+    count_bits(&local_count, bit_mask, grid_size, block_size);
+  }
+  cudaProfilerStop();
+  auto end = std::chrono::system_clock::now();
+  std::chrono::duration<double> elapsed_seconds = end-start;
+  std::cout << "Elapsed time (ms): " << elapsed_seconds.count()*1000 << std::endl;
+
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::destroy_bit_mask(bits));
+  free(local_valid);
+}
+
+TEST_F(BitMaskTest, MultiThreadedTest)
+{
+  const int num_rows = 100000;
+  bit_mask_t *bits = nullptr;
+ 
+  EXPECT_EQ(GDF_SUCCESS, bit_mask::create_bit_mask(&bits, num_rows, 0));
+
+  BitMask bit_mask(bits, num_rows);
+
+  test_unsafe_set_clear_g<<<1,bit_mask::detail::BITS_PER_ELEMENT>>>(bit_mask);
+
+  gdf_size_type local_count = 0;
+  EXPECT_EQ(GDF_SUCCESS, count_bits(&local_count, bit_mask));
+
+  if ((num_rows / 2) != local_count) {
+    std::cout << "  unsafe version got wrong count due to race condition" << std::endl;
+  } else {
+    std::cout << "  unsafe version got correct answer, race condition not triggered" << std::endl;
+  }
+  
```

Comment:
I think you need an expectation here, not just a print statement. Can you deterministically fail in the unsafe case?

Reply:
I can't guarantee failure... but the way I constructed the test, the odds of the unsafe version getting the correct answer are extremely low.  I will change it to an expect.